Mathlib/Combinatorics/SetFamily/AhlswedeZhang.lean

```diff
@@ -0,0 +1,430 @@
+
+@[simp] lemma truncatedSup_singleton (b a : α) : truncatedSup {b} a = if a ≤ b then b else ⊤ := by
+  simp [truncatedSup]; split_ifs <;> simp [*]
+
+lemma le_truncatedSup : a ≤ truncatedSup s a := by
+  rw [truncatedSup]
+  split_ifs with h
+  · obtain ⟨ℬ, hb, h⟩ := h
+    exact h.trans $ le_sup' id $ mem_filter.2 ⟨hb, h⟩
+  · exact le_top
+
+lemma map_truncatedSup (e : α ≃o β) (s : Finset α) (a : α) :
+    e (truncatedSup s a) = truncatedSup (s.map e.toEquiv.toEmbedding) (e a) := by
+  have : e a ∈ lowerClosure (s.map e.toEquiv.toEmbedding : Set β) ↔ a ∈ lowerClosure s
+  · simp
+  simp_rw [truncatedSup, apply_dite e, map_finset_sup', map_top, this]
+  congr with h
+  simp only [filter_map, Function.comp, Equiv.coe_toEmbedding, RelIso.coe_fn_toEquiv,
+    OrderIso.le_iff_le, id.def]
+  rw [sup'_map]
+  -- TODO: Why can't `simp` use `Finset.sup'_map`?
+  simp only [sup'_map, Equiv.coe_toEmbedding, RelIso.coe_fn_toEquiv, Function.comp_apply]
+
+variable [DecidableEq α]
+
+private lemma lower_aux : a ∈ lowerClosure ↑(s ∪ t) ↔ a ∈ lowerClosure s ∨ a ∈ lowerClosure t := by
+  rw [coe_union, lowerClosure_union, LowerSet.mem_sup_iff]
+
+lemma truncatedSup_union (hs : a ∈ lowerClosure s) (ht : a ∈ lowerClosure t) :
+    truncatedSup (s ∪ t) a = truncatedSup s a ⊔ truncatedSup t a := by
+  simpa only [truncatedSup_of_mem, hs, ht, lower_aux.2 (Or.inl hs), filter_union] using
+    sup'_union _ _ _
+
+lemma truncatedSup_union_left (hs : a ∈ lowerClosure s) (ht : a ∉ lowerClosure t) :
+    truncatedSup (s ∪ t) a = truncatedSup s a := by
+  simp only [mem_lowerClosure, mem_coe, exists_prop, not_exists, not_and] at ht
+  simp only [truncatedSup_of_mem, hs, filter_union, filter_false_of_mem ht, union_empty,
+    lower_aux.2 (Or.inl hs), ht]
+
+lemma truncatedSup_union_right (hs : a ∉ lowerClosure s) (ht : a ∈ lowerClosure t) :
+    truncatedSup (s ∪ t) a = truncatedSup t a := by rw [union_comm, truncatedSup_union_left ht hs]
+
+lemma truncatedSup_union_of_not_mem (hs : a ∉ lowerClosure s) (ht : a ∉ lowerClosure t) :
+    truncatedSup (s ∪ t) a = ⊤ := truncatedSup_of_not_mem fun h ↦ (lower_aux.1 h).elim hs ht
+
+lemma truncatedSup_of_isAntichain (hs : IsAntichain (· ≤ ·) (s : Set α)) (ha : a ∈ s) :
+    truncatedSup s a = a := by
+  refine' le_antisymm _ le_truncatedSup
+  simp_rw [truncatedSup_of_mem (subset_lowerClosure ha), sup'_le_iff, mem_filter]
+  rintro b ⟨hb, hab⟩
+  exact (hs.eq ha hb hab).ge
+
+end SemilatticeSup
+
+section SemilatticeInf
+variable [SemilatticeInf α] [BoundedOrder α] [@DecidableRel α (· ≤ ·)] [SemilatticeInf β]
+  [BoundedOrder β] [@DecidableRel β (· ≤ ·)] {s t : Finset α} {a : α}
+
+private lemma inf_aux : a ∈ upperClosure s → (s.filter (· ≤ a)).Nonempty :=
+  fun ⟨b, hb, hab⟩ ↦ ⟨b, mem_filter.2 ⟨hb, hab⟩⟩
+
+/-- The infimum of the elements of `s` less than `a` if there are some, otherwise `⊥`. -/
+def truncatedInf (s : Finset α) (a : α) : α :=
+  if h : a ∈ upperClosure s then (s.filter (· ≤ a)).inf' (inf_aux h) id else ⊥
```

Comment:
Can you explain why you add `⊥`? The natural choice would seem to be `top`, since that's the inf of an empty set.

Reply:
If the intent is not to actually forward-port it, we should just revert it in mathlib3 to get it off the dashboard

Reply:
I'm happy to make this a forward-port. But note that the second half of the file is not a forward-port

Reply:
In fact, it doesn't show up on the dashboard right now. So I'd be in favor of waiting for this PR to be merged before adding the aligns.